File: lib/js/test/src/test/TestHandler.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package test;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import org.apache.thrift.TException;

import thrift.test.Insanity;
import thrift.test.ThriftTest;
import thrift.test.Xception;
import thrift.test.Xception2;
import thrift.test.Xtruct;
import thrift.test.Xtruct2;
import thrift.test.Numberz;

public class TestHandler implements ThriftTest.Iface {

    public byte testByte(byte thing) throws TException {
        return thing;
    }

    public double testDouble(double thing) throws TException {
        return thing;
    }

    public Numberz testEnum(Numberz thing) throws TException {
        return thing;
    }

    public void testException(String arg) throws Xception, TException {
        throw new Xception(1,arg);       
    }

    public int testI32(int thing) throws TException {
        return thing;
    }

    public long testI64(long thing) throws TException {
        return thing;
    }

    public Map<Long, Map<Numberz, Insanity>> testInsanity(Insanity argument) throws TException {
        Map<Long, Map<Numberz, Insanity>> result = new HashMap<Long, Map<Numberz,Insanity>>();
        
        result.put(Long.valueOf(1), new HashMap<Numberz,Insanity>());
        result.get(Long.valueOf(1)).put(Numberz.ONE, argument);
        
        result.put(Long.valueOf(2), new HashMap<Numberz,Insanity>());
        result.get(Long.valueOf(2)).put(Numberz.ONE, argument);
        
        return result;
    }

    public List<Integer> testList(List<Integer> thing) throws TException {
        return thing;
    }

    public Map<Integer, Integer> testMap(Map<Integer, Integer> thing) throws TException {
        return thing;
    }

    public Map<Integer, Map<Integer, Integer>> testMapMap(int hello) throws TException {
        Map<Integer, Map<Integer,Integer>> result = new HashMap<Integer, Map<Integer,Integer>>();
        
        result.put(Integer.valueOf(1), new HashMap<Integer,Integer>());
        result.get(Integer.valueOf(1)).put(Integer.valueOf(1), Integer.valueOf(1));
        result.get(Integer.valueOf(1)).put(Integer.valueOf(2), Integer.valueOf(2));
        result.get(Integer.valueOf(2)).put(Integer.valueOf(1), Integer.valueOf(1));
    
        return result;
    }

    public Xtruct testMulti(byte arg0, int arg1, long arg2, Map<Short, String> arg3, Numberz arg4, long arg5) throws TException {
        Xtruct xtr = new Xtruct();
        
        xtr.byte_thing = arg0;
        xtr.i32_thing  = arg1;
        xtr.i64_thing  = arg2;
        xtr.string_thing = "server string";
        
        return xtr;
    }

    public Xtruct testMultiException(String arg0, String arg1) throws Xception, Xception2, TException {
        Xtruct xtr = new Xtruct();
        xtr.setString_thing(arg0);
        throw new Xception2(1,xtr);
    }

    public Xtruct2 testNest(Xtruct2 thing) throws TException {
       return thing;
    }

    public void testOneway(int secondsToSleep) throws TException {
        try{
            Thread.sleep(secondsToSleep * 1000);
        }catch(InterruptedException e){
            
        }
    }

    public Set<Integer> testSet(Set<Integer> thing) throws TException {
        return thing;
    }

    public String testString(String thing) throws TException {
        return thing;
    }

    public Xtruct testStruct(Xtruct thing) throws TException {
        return thing;
    }

    public long testTypedef(long thing) throws TException {
        return thing;
    }

    public void testVoid() throws TException {
        
    }
}

File: lib/csharp/test/TestStreamServer/TestHandler.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Thrift.Server;
using Thrift.Collections;

namespace Thrift.Test
{
    public class TestHandler : ThriftTest.Iface
    {
        public TServer server;

        public TestHandler() { }

        public void testVoid()
        {

        }

        public string testString(string thing)
        {
            return thing;
        }

        public sbyte testByte(sbyte thing)
        {
            return thing;
        }

        public int testI32(int thing)
        {
            return thing;
        }

        public long testI64(long thing)
        {
            return thing;
        }

        public double testDouble(double thing)
        {
            return thing;
        }

        public byte[] testBinary(byte[] thing)
        {
            string hex = BitConverter.ToString(thing).Replace("-", string.Empty);
            return thing;
        }

        public Xtruct testStruct(Xtruct thing)
        {
            return thing;
        }

        public Xtruct2 testNest(Xtruct2 nest)
        {
            Xtruct thing = nest.Struct_thing;
            return nest;
        }

        public Dictionary<int, int> testMap(Dictionary<int, int> thing)
        {
            bool first = true;
            foreach (int key in thing.Keys)
            {
                if (first)
                {
                    first = false;
                }
                else
                {
                }
            }
            return thing;
        }

        public Dictionary<string, string> testStringMap(Dictionary<string, string> thing)
        {
            bool first = true;
            foreach (string key in thing.Keys)
            {
                if (first)
                {
                    first = false;
                }
                else
                {
                }
            }
            return thing;
        }

        public THashSet<int> testSet(THashSet<int> thing)
        {
            bool first = true;
            foreach (int elem in thing)
            {
                if (first)
                {
                    first = false;
                }
                else
                {
                }
            }
            return thing;
        }

        public List<int> testList(List<int> thing)
        {
            bool first = true;
            foreach (int elem in thing)
            {
                if (first)
                {
                    first = false;
                }
                else
                {
                }
            }
            return thing;
        }

        public Numberz testEnum(Numberz thing)
        {
            return thing;
        }

        public long testTypedef(long thing)
        {
            return thing;
        }

        public Dictionary<int, Dictionary<int, int>> testMapMap(int hello)
        {
            Dictionary<int, Dictionary<int, int>> mapmap =
              new Dictionary<int, Dictionary<int, int>>();

            Dictionary<int, int> pos = new Dictionary<int, int>();
            Dictionary<int, int> neg = new Dictionary<int, int>();
            for (int i = 1; i < 5; i++)
            {
                pos[i] = i;
                neg[-i] = -i;
            }

            mapmap[4] = pos;
            mapmap[-4] = neg;

            return mapmap;
        }

        public Dictionary<long, Dictionary<Numberz, Insanity>> testInsanity(Insanity argument)
        {

            Xtruct hello = new Xtruct();
            hello.String_thing = "Hello2";
            hello.Byte_thing = 2;
            hello.I32_thing = 2;
            hello.I64_thing = 2;

            Xtruct goodbye = new Xtruct();
            goodbye.String_thing = "Goodbye4";
            goodbye.Byte_thing = (sbyte)4;
            goodbye.I32_thing = 4;
            goodbye.I64_thing = (long)4;

            Insanity crazy = new Insanity();
            crazy.UserMap = new Dictionary<Numberz, long>();
            crazy.UserMap[Numberz.EIGHT] = (long)8;
            crazy.Xtructs = new List<Xtruct>();
            crazy.Xtructs.Add(goodbye);

            Insanity looney = new Insanity();
            crazy.UserMap[Numberz.FIVE] = (long)5;
            crazy.Xtructs.Add(hello);

            Dictionary<Numberz, Insanity> first_map = new Dictionary<Numberz, Insanity>();
            Dictionary<Numberz, Insanity> second_map = new Dictionary<Numberz, Insanity>(); ;

            first_map[Numberz.TWO] = crazy;
            first_map[Numberz.THREE] = crazy;

            second_map[Numberz.SIX] = looney;

            Dictionary<long, Dictionary<Numberz, Insanity>> insane =
              new Dictionary<long, Dictionary<Numberz, Insanity>>();
            insane[(long)1] = first_map;
            insane[(long)2] = second_map;

            return insane;
        }

        public Xtruct testMulti(sbyte arg0, int arg1, long arg2, Dictionary<short, string> arg3, Numberz arg4, long arg5)
        {
            Xtruct hello = new Xtruct(); ;
            hello.String_thing = "Hello2";
            hello.Byte_thing = arg0;
            hello.I32_thing = arg1;
            hello.I64_thing = arg2;
            return hello;
        }

        public void testException(string arg)
        {
            if (arg == "Xception")
            {
                Xception x = new Xception();
                x.ErrorCode = 1001;
                x.Message = arg;
                throw x;
            }
            if (arg == "TException")
            {
                throw new Thrift.TException("TException");
            }
            return;
        }

        public Xtruct testMultiException(string arg0, string arg1)
        {
            if (arg0 == "Xception")
            {
                Xception x = new Xception();
                x.ErrorCode = 1001;
                x.Message = "This is an Xception";
                throw x;
            }
            else if (arg0 == "Xception2")
            {
                Xception2 x = new Xception2();
                x.ErrorCode = 2002;
                x.Struct_thing = new Xtruct();
                x.Struct_thing.String_thing = "This is an Xception2";
                throw x;
            }

            Xtruct result = new Xtruct();
            result.String_thing = arg1;
            return result;
        }

        public void testStop()
        {
            if (server != null)
            {
                server.Stop();
            }
        }

        public void testOneway(int arg)
        {
            System.Threading.Thread.Sleep(arg * 1000);
        }
    }

}
